include/votca/xtp/eigencuda.h

```diff
@@ -0,0 +1,131 @@
+#include <curand.h>
+#include <sstream>
+#include <vector>
+#include <votca/xtp/eigen.h>
+
+/*
+ * \brief Perform matrix-matrix multiplication in a GPU
+ *
+ * The `EigenCuda` class handles the allocation and deallocation of arrays on
+ * the GPU. Firstly, to perform a matrix multiplication, memory must be
+ * allocated in the device to contain the involved matrices. The
+ * `initialize_matrix_mem` method firstly allocates memory by calling the
+ * `gpu_alloc` method that allocates either pinned or pageable memory, see:
+ * https://devblogs.nvidia.com/how-optimize-data-transfers-cuda-cc/ Then the
+ * array could be optionally copy to the device.
+ */
+
+namespace votca {
+namespace xtp {
+
+inline cudaError_t checkCuda(cudaError_t result) {
+// Check Cuda error
+#if defined(DEBUG) || defined(_DEBUG)
+  if (result != cudaSuccess) {
+    std::cerr << "CUDA Runtime Error: " << cudaGetErrorString(result) << "\n";
+  }
+#endif
+  return result;
+}
+
+// Structure with the sizes to call ?GEMM
+struct Shapes {
+  int A_rows;
+  int A_cols;
+  int B_rows;
+  int B_cols;
+  int C_rows;
+
+  Shapes(long int _a_rows, long int _a_cols, long int _b_rows, long int _b_cols,
+         long int _c_rows)
+      : A_rows{static_cast<int>(_a_rows)},
+        A_cols{static_cast<int>(_a_cols)},
+        B_rows{static_cast<int>(_b_rows)},
+        B_cols{static_cast<int>(_b_cols)},
+        C_rows{static_cast<int>(_c_rows)} {}
+};
+
+// col Major for CUDA
+template <typename T>
+using Mat = Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic, Eigen::ColMajor>;
+
+template <typename T>
+class EigenCuda {
+
+ public:
+  EigenCuda() {
+    cublasCreate(&_handle);
+    cudaStreamCreate(&_stream);
+  }
+  EigenCuda(bool pinned) : _pinned{pinned} { EigenCuda{}; }
+
+  // Deallocate both the handler and allocated arrays
+  ~EigenCuda();
+
+  // Remove the copy operations
+  EigenCuda(const EigenCuda &) = delete;
+  EigenCuda &operator=(const EigenCuda &) = delete;
+
+  // Perform the triple matrix multiplication A * matrix * C, for the vector
+  // of matrices given by tensor
+  std::vector<Mat<T>> triple_tensor_product(const Mat<T> &A, const Mat<T> &C,
+                                            const std::vector<Mat<T>> &tensor);
+
+  // Perform a multiplication between a matrix and a tensor
+  std::vector<Mat<T>> right_matrix_tensor(
+      const Mat<T> &A, const std::vector<Mat<T>> &tensor) const;
+
+ private:
+  // Check available memory
+  void check_available_memory(size_t required) const;
+
+  // Allocate memory in the device
+  void gpu_alloc(T **x, std::size_t n) const;
+
+  // Deallocate memory from the device
+  void gpu_free(T *x) const;
+
+  // Allocate matrix in the device
+  T *initialize_matrix_mem(size_t size_A) const;
+
+  // Allocate memory for a matrix and copy it to the device
+  T *initialize_and_copy(const Mat<T> &A) const;
+
+  // Invoke the ?gemm function of cublas
+  void gemm(Shapes shapes, const T *dA, const T *dB, T *dC) const;
```

Comment:
I have removed the templates. I have call the pointer `dA`, `dB` and `dC` because in the CUDA community people use the first letter `d` to denote that is a pointer in the device